File: tracks/json_from_kmz.pl
#!/usr/bin/env perl
use strict;
use warnings;
use JSON ();
use File::Basename qw(basename);
use File::Spec::Functions qw(catfile catdir);
use Date::Parse ();
use lib catfile($ENV{HOME}, 'bin');
use Misc qw(dirname assert note warning error fatal internal);
sub do_dir($$);
sub get_info_from_kmz($);
sub encode_json($);
sub decode_json($);

my $all_info = [];
my $info_json = decode_json(Misc::get('info.json'));
for my $dir (sort keys %$info_json) {
  my $extra = $info_json->{$dir};
  push(@$all_info, { dir => $dir, name => $extra->{name}, data => do_dir($dir, $extra) });
}
print "main.json\n";
Misc::put('main.json', encode_json($all_info));
exit;

#my $all_info = [];
#for my $dir (grep(-d, sort(glob('*')))) {
#  my $dir_info = do_dir($dir);
#  if (defined($dir_info)) {
#    push(@$all_info, { dir => $dir, data => $dir_info });
#  }
#}
#print "main.json\n";
#Misc::put('main.json', encode_json($all_info));

exit;

sub do_dir($$) {
  my($dir, $extra) = @_;
  my $dir_info = [];
  my @kmzs = glob("$dir/*.kmz");
  return undef unless @kmzs;
  for my $kmz (@kmzs) {
    my $info = get_info_from_kmz($kmz);
    my $json = Misc::subst($kmz, ' *\.kmz$' => '.json', '/ ' => '/', ' +' => '_');
    print "$json\n";
    if (defined(my $e = $extra->{basename($json)})) {
      for my $key (keys %$e) {
        $info->{$key} = $e->{$key};
      }
    }
    Misc::put($json, encode_json($info));
    push(@$dir_info,
      { path => $json, map { $_ => $info->{$_} } qw(date name Distance StartCoord snippet) });
  }
  return [ sort { $b->{date} cmp $a->{date} } @$dir_info ];
  #return $dir_info;
}

sub get_info_from_kmz($) {
  my($kmz) = @_;
  my $xml = do {
    local $/ = undef;
    my $fh = Misc::do_open('-|', "gunzip < '$kmz'");
    <$fh>;
  };
  my %info = ();
  while ($xml =~ m{<Placemark>(.*?)</Placemark>}gs) {
    my $x = $1;
    if ($x =~ m{<Snippet>(.*?)</Snippet>}) {
      $info{snippet} = Misc::subst($1, '.*? [ap]m\b ?' => '');
      if ($x =~ m{<name>(.*?)</name>}) {
        $info{name} = $1;
      }
      while ($x =~ m{<tr>\s*<td[^<>]*>(.*?):?</td>\s*<td[^<>]*>(.*?)</td>}gs) {
        my($key, $value) = ($1, $2);
        $info{$key} = $value;
      }
    }
    if (my $date = $info{Date}) {
      if ($date =~ s/^(.*?)  *(\d+:\d+ [ap]m)$//) {
        $info{Date} = $1;
        $info{Time} = $2;
      }
    }
  }
  my(undef,undef,undef,$day,$month,$year) = Date::Parse::strptime($info{'Start Time'});
  $info{date} = sprintf('%04d/%02d/%02d', $year+1900, $month+1, $day);

  my @coordinates = ();
  if ($xml =~ m{.*<coordinates>\s*(.*?)\s*</coordinates>}s) {
    my $coordinates = $1;
    my($lat0, $lng0) = (0, 0);
    while ($coordinates =~ /([-.\d]+),([-.\d]+),/g) {
      my($lng, $lat) = ($1, $2);
      if ($lat ne $lat0 || $lng ne $lng0) {
        push(@coordinates, { lat => $lat, lng => $lng });
        $lat0 = $lat;
        $lng0 = $lng;
      }
    }
    $info{StartCoord} = $coordinates[0];
  }
  $info{coords} = \@coordinates;
  return \%info;
}

sub encode_json($) {
  my($data) = @_;
  my $j = JSON->new->indent->space_after->encode($data);
  $j =~ s/\{\s*"lat": *"(.*?)",\s*"lng":\s*"(.*?)"\s*\}/{"lat": $1, "lng": $2}/g;
  $j =~ s/\{\s*"lng": *"(.*?)",\s*"lat":\s*"(.*?)"\s*\}/{"lat": $2, "lng": $1}/g;
  $j =~ s/^   / /gm;
  $j =~ s/^    /  /gm;
  return $j;
}

sub decode_json($) {
  my($json) = @_;
  return JSON->new->decode($json);
}

File: tracks/json_from_kmz.py
#!/usr/bin/env python
import collections
import json
from glob import glob
import re
import os
import xml.etree.ElementTree as ElementTree
import dateutil.parser
from zipfile import ZipFile, BadZipfile

def main():
  info_json = get_json('info.json')
  all_info = map(lambda dir: do_dir(dir, info_json[dir]), info_json)
  put_json('main.json', all_info)

def do_dir(dir, extra):
  dir_info = [ do_file(kmz_path, extra) for kmz_path in glob('%s/*.kmz' % dir) ]
  return { 'dir': dir, 'name': extra['name'], 'data': sorted(dir_info, key=lambda x: x['date']) }

def do_file(kmz_path, extra):
  info = get_info_from_kmz(kmz_path)
  json_path = sub(kmz_path, { r' +': '_', r'/ +': '/', r' *\.kmz$': '.json' })
  base = os.path.basename(json_path)
  if base in extra:
    info = dict(info.items() + extra[base].items())
  put_json(json_path, info)
  return dict(
    { key: info[key] for key in ['date', 'name', 'Distance', 'StartCoord', 'snippet'] },
    path=json_path)

def get_info_from_kmz(kmz):
  print 'kmz:', kmz
  try:
    kml = ZipFile(kmz, 'r').read('doc.kml')
    ns = '{http://earth.google.com/kml/2.2}'
    # the Placemark containing Snippet has the info we want
    placemark = ElementTree.fromstring(kml).find('.//{0}Snippet/..'.format(ns))
    coordinates = list(extract_coordinates(
      placemark.find('./{0}MultiGeometry/{0}LineString/{0}coordinates'.format(ns)).text))
    info = dict((re.sub(r':$', '', tr[0].text), tr[1].text)
        for tr in placemark.find(ns + 'description').iter(ns + 'tr'))
    start = dateutil.parser.parse(info['Start Time'])
    return dict(
      info,
      name=placemark.find(ns + 'name').text,
      snippet=re.sub(r'.*? [ap]m\b ?', '', placemark.find(ns + 'Snippet').text),
      date=str(start.date()),
      time=str(start.time()),
      coords=coordinates,
      StartCoord=coordinates[0],
    )
  except BadZipfile as e:
    print 'Error reading "{0}": {1}'.format(kmz, e)

# Extract the "<lat>,<lng>" coordinates from text, skipping duplicates.
def extract_coordinates(text):
  prev = { 'lng': 0, 'lat': 0 }
  for match in re.compile(r'([-.\d]+),([-.\d]+),').finditer(text):
    c = { 'lng': float(match.group(1)), 'lat': float(match.group(2)) }
    if c != prev:
      prev = c
      yield c

def get_json(path):
  with open('info.json') as data_file:    
    return json.load(data_file)

# Write data as json to path. It is pretty-printed with lat/lng done compactly.
def put_json(path, data):
  with open(path, 'w') as outfile:
    outfile.write(re.sub(
      r'{\s*("lat": *\S+),\s*("lng": *\S+)\s*}', r'{ \1, \2 }',
      json.dumps(data, indent=1, sort_keys=True)))

# Substitute key -> value from subs in string
def sub(string, subs):
  assert isinstance(subs, collections.Mapping), 'Expected dict, got: {0}'.format(subs)
  for key, value in subs.items():
    string = re.sub(key, value, string)
  return string

main()
